src/test/java/echoserver/EchoServerTest.java

```diff
@@ -2,50 +2,108 @@
 
 import org.junit.jupiter.api.Test;
 
-import java.io.*;
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
 
 import static org.junit.jupiter.api.Assertions.*;
 
 class EchoServerTest {
     @Test
-    void itEchosText() throws IOException {
-        EchoServer echoServer = new EchoServer();
-        ArrayList<String> input = new ArrayList<String>(Arrays.asList("This", "Should", "be", "echoed"));
-        TestConnection connection = new TestConnection(input);
-        echoServer.echo(connection);
+    void itConnectsToClientsInOrderReceived() throws IOException {
+        FakeEchoClient echoClient = new FakeEchoClient();
+        ClientSocketConnection first_client = new ClientSocketConnection();
+        ClientSocketConnection second_client = new ClientSocketConnection();
+        ClientSocketConnection third_client = new ClientSocketConnection();
+        ArrayList<ClientSocketConnection> clients = new ArrayList<ClientSocketConnection>(Arrays.asList(first_client, second_client, third_client));
+        FakeListener listener = new FakeListener(clients);
+        EchoServer echoServer = new EchoServer(listener, echoClient);
 
-        assertEquals(Arrays.asList("This", "Should", "be", "echoed"), connection.written);
+        echoServer.connect();
+
+        assertEquals(Arrays.asList(first_client, second_client, third_client),listener.connectedClients);
     }
 
-    private class TestConnection implements Connection {
+    @Test
+    void itClosesTheConnection() throws IOException {
+        FakeEchoClient echoClient = new FakeEchoClient();
+        ExceptionalListener listener = new ExceptionalListener();
+        EchoServer echoServer = new EchoServer(listener, echoClient);
 
-        public ArrayList<String> written;
-        private ArrayList<String> input;
+        echoServer.connect();
 
-        public TestConnection(ArrayList<String> input) {
-            this.input = input;
-            this.written = new ArrayList<String>();
-        }
+        assertTrue(listener.closed);
+    }
+
+    private class ClientSocketConnection implements Connection {
```

Comment:
Is there something specific about this double that warrants this name?

FWIW, you could have named it `TestConnection` and it would not collide with the existing `TestConnection`. Since you defined them as inner classes, they are distinct from each other.

Reply:
Will make the change